samples/csv-chatbot/src/main/java/io/quarkiverse/langchain4j/sample/chatbot/MovieDatabaseContentRetriever.java

```diff
@@ -0,0 +1,61 @@
+package io.quarkiverse.langchain4j.sample.chatbot;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.sql.DataSource;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+import dev.langchain4j.rag.content.Content;
+import dev.langchain4j.rag.content.retriever.ContentRetriever;
+import dev.langchain4j.rag.query.Query;
+import io.quarkus.logging.Log;
+import io.vertx.core.json.JsonObject;
+
+@ApplicationScoped
+public class MovieDatabaseContentRetriever implements ContentRetriever {
+
+    @Inject
+    MovieMuseSupport support;
+
+    @Inject
+    DataSource dataSource;
+
+    @Override
+    public List<Content> retrieve(Query query) {
+        String question = query.text();
+        String sqlQuery = support.createSqlQuery(question);
+        if (sqlQuery.contains("```sql")) { // strip the formatting if it's there
+            sqlQuery = sqlQuery.substring(sqlQuery.indexOf("```sql") + 6, sqlQuery.lastIndexOf("```"));
+        }
+        Log.info("Question to answer: " + question);
+        Log.info("Supporting SQL query: " + sqlQuery);
+        List<Content> results = new ArrayList<>();
+        Log.info("Retrieved relevant movie data: ");
+        try (Connection connection = dataSource.getConnection()) {
+            try (Statement statement = connection.createStatement()) {
+                try (ResultSet resultSet = statement.executeQuery(sqlQuery)) {
+                    while (resultSet.next()) {
+                        JsonObject json = new JsonObject();
+                        int columnCount = resultSet.getMetaData().getColumnCount();
+                        for (int i = 1; i <= columnCount; i++) {
+                            String columnName = resultSet.getMetaData().getColumnName(i);
+                            json.put(columnName, resultSet.getObject(i));
+                        }
+                        Log.info("Item: " + json);
+                        results.add(Content.from(json.toString()));
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+        return results;
+    }
+}
```

Comment:
Can't we use the `StatelessSession` here? It should cut down on the code

Reply:
I don't think so, the problem is that we don't have that much control over the SQL query that the LLM creates.  We don't know which columns of the table will be in the SELECT clause and the result potentially might not be mappable via ORM to a `Movie` instance at all.

Reply:
So instead of mapping it via ORM, we just take the whole ResultSet, along with column names from its metadata, and transform it blindly into a JSON object.

Reply:
Gotcha